lessons/snake/logic/index.md

```diff
@@ -0,0 +1,998 @@
+        new_y = old_y + dir_y
+```
+
+Směr hada teď můžeš měnit změnou `snake_direction` v `__init__`.
+Funguje to? (Jestli ne, oprav to – a jestli to nejde, zavolej někoho na pomoc!)
+
+Nyní zbývá jen atribut `snake_direction` měnit, když uživatel něco stiskne na
+klávesnici.
+To už je doména Pygletu.
+Opusť na chvíli abstraktní stav v `had.py` a koukni na hru v `ui.py`.
+Sem je potřeba přidat funkci, která reaguje na stisk klávesy.
+
+Aby Pyglet tuhle funkci našel a uměl zavolat, musí se jmenovat `on_key_press`,
+musí mít dekorátor `@window.event`, a musí brát dva argumenty:
+číslo klávesy, která byla zmáčknutá, a informace o modifikátorech
+(Shift, Ctrl, a tak podobně):
+
+```python
+@window.event
+def on_key_press(symbol, mod):
+    ...
+```
+
+Druhý argument nepoužijeme.
+Podle prvního ale nastav aktuální směr hada.
+Čísla kláves jsou definována v modulu `pyglet.window.key` jako konstanty se
+jmény `LEFT`, `ENTER`, `Q` či `AMPERSAND` .
+My použijeme šipky:
+
+```python
+@window.event
+def on_key_press(symbol, mod):
+    if symbol == pyglet.window.key.LEFT:
+        state.snake_direction = -1, 0
+    if symbol == pyglet.window.key.RIGHT:
+        state.snake_direction = 1, 0
+    if symbol == pyglet.window.key.DOWN:
+        state.snake_direction = 0, -1
+    if symbol == pyglet.window.key.UP:
+        state.snake_direction = 0, 1
+```
+
+Tuhle funkci je potřeba dát někam za nastavení `window` (aby byl k dispozici
+`window.event`) a před `pyglet.app.run()` (protože nastavovat ovládání až
+potom, co hra proběhne, je zbytečné).
+Nejlepší je ji dát vedle jiné funkce s dekorátorem `@window.event`,
+aby byly pěkně pohromadě.
+
+Funguje to?
+Můžeš ovládat směr hada?
+To je skvělé!
+Určitě ale při zkoušení narazíš na pár věcí, které je potřeba dodělat:
+
+* Had by neměl mít možnost vylézt ven z okýnka.
+* Had by měl jíst jídlo a růst.
+* Hra by měla skončit, když had narazí sám do sebe.
+
+Pojďme je vyřešit, jednu po druhé.
+
+
+## Zatím dobrý, teď ale narazíme
+
+„Hadí“ hry jako ta naše mají dvě varianty: buď je kolem hřiště „zeď“
+a hráč při nárazu do okraje prohraje, nebo je hřiště „nekonečné“ – had okrajem
+proleze a objeví se na druhé straně.
+
+My nakonec naprogramujeme druhou variantu, která je zajímavější.
+Začneme ale s tou první, která je jednodušší.
+
+Vrať se k souboru se stavem – `had.py`.
+Budeme pracovat na chování, na logice hry; ne na vykreslování a ovládání.
+
+Abys zjistil{{a}}, jestli had „vylezl“ z levého okraje okna ven,
+je potřeba zkontrolovat, jestli <var>x</var>-ová souřadnice hlavy
+je menší než 0.
+To je dobré udělat hned poté, co nové souřadnice hlavy získáš – konkrétně
+hned před řádkem `new_head = new_x, new_y` v metodě `move`.
+
+A co při takovém nárazu udělat?
+Určitě sis už všiml{{a}}, že jakákoli chyba ukončí program.
+Nejjednodušší způsob ukončení programu je právě to – stačí tedy vyvolat chybu.
+Nejkratší způsob, jak v Pythonu vyvolat na určitém místě chybu, je dělení
+nulou: `1/0`.
```

Comment:
Tohle jsem viděl a trochu se zhrozil. Nebylo by lepší použít `sys.exit()`. Já vím, že nakonec to bude pryč a dělení nulou tam nezůstane, ale stejně mi tento způsob vysvětlování přijde prapodivný.

Reply:
Taky se zpětně divím, co mě to napadlo. Kaju se. Na workshopu byl `exit()`.

Reply:
Hmm, `1/0` je ještě lepší než `assert 0`, díky :smile:

Reply:
![stop](https://user-images.githubusercontent.com/2401856/37904706-e2fc4f96-30fc-11e8-9cbb-5c19360e8082.jpg)

Reply:
https://github.com/pyvec/naucse.python.cz/pull/357